lib/Dialect/LLHD/IR/LLHDOps.cpp

```diff
@@ -663,22 +663,30 @@ static LogicalResult verify(llhd::EntityOp op) {
                "Cannot have more inputs than arguments, expected at most ")
            << numArgs << " but got: " << nIns;
   }
+
+  // Check that all block arguments are of signal type
+  for (size_t i = 0; i < numArgs; ++i)
+    if (!op.getArgument(i).getType().isa<llhd::SigType>())
+      return op.emitError("usage of invalid argument type. Got ")
+             << op.getArgument(i).getType() << ", expected LLHD signal type";
+
   return success();
 }
 
 LogicalResult circt::llhd::EntityOp::verifyType() {
+  FunctionType type = getType();
+
   // Fail if function returns any values. An entity's outputs are specially
   // marked arguments.
-  if (getNumResults() > 0)
+  if (type.getNumResults() > 0)
     return emitOpError("an entity cannot have return types.");
 
   // Check that all operands are of signal type
-  for (int i = 0, e = getNumFuncArguments(); i < e; ++i) {
-    if (!getArgument(i).getType().isa<llhd::SigType>()) {
+  for (Type inputType : type.getInputs())
+    if (!inputType.isa<llhd::SigType>())
       return emitOpError("usage of invalid argument type. Got ")
-             << getArgument(i).getType() << ", expected LLHD signal type";
-    }
-  }
+             << inputType << ", expected LLHD signal type";
+
```

Comment:
Much cleaner :+1:. I still wonder if at some point we'd want to change LLHD entities to not partition inputs from outputs by a fixed offset, but rather just having one plain argument list with directionality annotated as an attribute. That could make linking much easier, especially if we make port connections by name.

Reply:
That would make sense to me, and might make processing like this a little more simple. I'll leave this as future work for now.